src/main/java/net/snowflake/client/core/JsonSqlInput.java

```diff
@@ -187,47 +209,118 @@ public Timestamp readTimestamp(TimeZone tz) throws SQLException {
           if (value == null) {
             return null;
           }
-          int columnType = ColumnTypeHelper.getColumnType(fieldMetadata.getType(), session);
-          int columnSubType = fieldMetadata.getType();
-          int scale = fieldMetadata.getScale();
-          Timestamp result =
-              SqlInputTimestampUtil.getTimestampFromType(
-                  columnSubType, (String) value, session, sessionTimeZone, tz);
-          if (result != null) {
-            return result;
+          return formatTimestamp(tz, value, fieldMetadata);
+        });
+  }
+
+  @Override
+  public <T> T readObject(Class<T> type) throws SQLException {
+    return withNextValue(
+        (value, jsonNode, fieldMetadata) -> {
+          if (SQLData.class.isAssignableFrom(type)) {
+            if (jsonNode.isNull()) {
+              return null;
+            } else {
+              SQLInput sqlInput =
+                  new JsonSqlInput(
+                      jsonNode, session, converters, fieldMetadata.getFields(), sessionTimeZone);
+              SQLData instance = (SQLData) SQLDataCreationHelper.create(type);
+              instance.readSQL(sqlInput, null);
+              return (T) instance;
+            }
+          } else if (Map.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            } else {
+              return (T) convertSqlInputToMap((SQLInput) value);
+            }
+          } else if (String.class.isAssignableFrom(type)
+              || Boolean.class.isAssignableFrom(type)
+              || Byte.class.isAssignableFrom(type)
+              || Short.class.isAssignableFrom(type)
+              || Integer.class.isAssignableFrom(type)
+              || Long.class.isAssignableFrom(type)
+              || Float.class.isAssignableFrom(type)
+              || Double.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            }
+            return (T) value;
+          } else if (Date.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            }
+            return (T) formatDate((String) value);
+          } else if (Time.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            }
+            return (T) formatTime((String) value);
+          } else if (Timestamp.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            }
+            return (T) formatTimestamp(sessionTimeZone, value, fieldMetadata);
+          } else if (byte[].class.isAssignableFrom(type)) {
```

Comment:
Does it work for `[]Byte.class` as well?

Reply:
No Our ByteConverter can return only byte[]. It couldn't be simply casted